imap_processing/spice/kernels.py

```diff
@@ -0,0 +1,158 @@
+
+    Parameters
+    ----------
+    f_py : Callable
+        The function requiring SPICE that we are going to wrap if being used
+        explicitly, otherwise None, in which case ensure_spice is being used,
+        not as a function wrapper (see l2a_processing.py) but as a true
+        decorator without an explicit function argument.
+    time_kernels_only : bool
+        Specify that we only need to furnish time kernels (if SPICE_METAKERNEL
+        is set, we still just furnish that metakernel and assume the time
+        kernels are included.
+
+    Returns
+    -------
+    Callable
+        Decorated function, with spice error handling.
+
+    Notes
+    -----
+    Before trying to understand this piece of code, read this:
+    https://stackoverflow.com/questions/5929107/decorators-with-parameters/60832711#60832711
+
+    **Control flow overview:**
+    1. Try simply calling the wrapped function naively.
+        * SUCCESS? Great! We're done.
+        * SpiceyError? Go to step 2.
+
+    2. Furnish metakernel at SPICE_METAKERNEL
+        * SUCCESS? Great, return the original function again (so it can be
+          re-run).
+        * KeyError? Seems like SPICE_METAKERNEL isn't set, no problem. Go to
+          step 3.
+
+    3. Did we get the parameter time_kernels_only=True?
+    --> YES? We only need LSK and SCLK kernels to run this function. Go fetch
+        those and furnish and return the original function (so it can be re-run).
+    --> NO? Dang. This is sort of the end of the line. Re-raise the error
+        generated from the failed spiceypy function call but add a better
+        message to it.
+
+    Examples
+    --------
+    There are three ways to use this object
+
+    1. A decorator with no arguments
+        >>> @ensure_spice
+        ... def my_spicey_func(a, b):
+        ...     pass
+
+    2. A decorator with parameters. This is useful
+    if we only need the latest SCLK and LSK kernels for the function involved.
+        >>> @ensure_spice(time_kernels_only=True)
+        ... def my_spicey_time_func(a, b):
+        ...     pass
+
+    3. An explicit wrapper function, providing a dynamically set value for
+    parameters, e.g. time_kernels_only
+        >>> wrapped = ensure_spice(spicey_func, time_kernels_only=True)
+        ... result = wrapped(*args, **kwargs)
+    """
+    if f_py and not callable(f_py):
+        raise ValueError(
+            f"Received a non-callable object {f_py} as the f_py argument to"
+            f"ensure_spice.  f_py must be a callable object."
+        )
+
+    def _decorator(func: Callable[..., Callable]) -> Callable:
+        """
+        Decorate or wrap input function depending on how ensure_spice is used.
+
+        Parameters
+        ----------
+        func : Callable
+            The function to be decorated/wrapped.
+
+        Returns
+        -------
+        Callable
+            If used as a function wrapper, the decorated function is returned.
+        """
+
```

Comment:
Sorry for the noise here. I see that I missed the docstring about two paths with keyword arguments or without. Resolving this, but if we don't need the `time_kernels_only` argument on IMAP, then it could be simplified to avoid so much nesting.

To me, the `time_kernels_only` seems like you might be doing too much in this one function and a separate function specifically for `retrieve_time_kernels()` could be warranted in that case to avoid any confusion about "ensuring spice" doing more than just ensuring, it would also be downloading.

Reply:
Do we need the`time_kernels_only` argument on IMAP? In l1a processing, we need to be able to convert SHCOARSE to J2000 seconds which will require us to furnish an sclk and leapsecond kernel. We could achieve that by furnishing a full metakernel that also provides all the kernels required for computing geometry but which will not be used in l1a. Doing it that way makes is slightly harder to document the SPICE kernels that were used in processing because unused kernels are being furnished. I can certainly see that we could write a separate decorator that is specific to functions that only require the time kernels. I don't know if it is preferable to have two double nested decorators instead of one triple nested decorator.

There are so many interconnected decisions on implementation that ultimately link to the decisions we make about how we will manage and use kernels. I need to present the two main options that I am aware of to the team so that you all can help with that decision. In the meantime, this decorator is meant to be a starting point that will allow us to develop functions that use SPICE without having to think about  the intricacies of furnishing required kernels.

Reply:
That all makes sense to me. I honestly just don't know enough about any of it, so your recommendations / suggestions are fantastic. I see your point about only needing the time kernels for some levels and only limiting the documentation to that. I was thinking that we were just documenting the "metakernel" in a sense and dumping everything. Does it also make it slower to load all the geometry/attitude kernels if they aren't used?

I also agree that this is a really nice starting point 👍

Reply:
Loading kernels has very little overhead. SPICE really just reads some metadata from the kernel that is needed to know which kernel to use when it comes time.

Documenting kernels is another interconnected decision to make. If we just want to furnish the metakernels (we will have one attitude metakernel and one ephemeris metakernel) and document that those were furnished, that works. It is possible to identify exactly which kernels affect each product and document just that subset of kernels too. On EMM, the SDC built the capability to generate a custom metakernel for a given time range that only included the specific kernels which provide coverage for that time range.